### PR TITLE
Add settings page and show quota usage and resource limits for the project

### PR DESCRIPTION
**Pull request by jwforres**

### CODE REVIEW

**Comment by jwforres** [comment]
@liggitt @derekwaynecarr

File: assets/app/views/settings.html [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+        <dt>Display name</dt>
+        <dd ng-if="project.displayName">{{project.displayName}}</dd>
+        <dd ng-if="!project.displayName"><em>No display name</em></dd>
+        <dt>Description</dt>
+        <dd ng-if="project.metadata.annotations && project.metadata.annotations.description">{{project.metadata.annotations.description}}</dd>
+        <dd ng-if="!project.metadata.annotations || !project.metadata.annotations.description"><em>No description</em></dd>
+      </dl>
+      <h2>
+        <span>Quota usage</span>
+        <a href="javascript:;" style="margin-left: 5px; cursor: pointer;" class="small pficon pficon-help" data-container="body" data-toggle="popover" data-placement="right" data-trigger="focus" data-content="Quota usage is aggregated for the whole project.  Once the maximum for the project has been reached no more of that resource can be allocated within the project."></a>
```
Comment (jwforres) [inline assets/app/views/settings.html]: @derekwaynecarr review this help message

File: assets/app/views/settings.html [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+            <td>
+              <span ng-if="!quota.status.hard">{{specMax  | usageWithUnits : resourceType}}</span>
+              <span ng-if="quota.status.hard">{{quota.status.hard[resourceType]  | usageWithUnits : resourceType}}</span>
+            </td>
+          </tr>
+        </tbody>
+      </table>
+      <h2>
+        <span>Resource limits</span>
+        <a href="javascript:;" style="margin-left: 5px; cursor: pointer;" class="small pficon pficon-help" data-container="body" data-toggle="popover" data-placement="right" data-trigger="focus" data-content="Resource limits define minimum and maximum values for resources that can be created within the project."></a>
```
Comment (jwforres) [inline assets/app/views/settings.html]: @derekwaynecarr review this help message

File: assets/app/views/settings.html (hunk 1 above)
Comment (derekwaynecarr) [inline assets/app/views/settings.html]: Does this render if and only if a quota is on for the project?

The message looks fine.

File: assets/app/views/settings.html (hunk 2 above)
Comment (derekwaynecarr) [inline assets/app/views/settings.html]: LGTM

**Comment by derekwaynecarr** [comment]
LGTM

**Comment by jwforres** [comment]
Also happens to fix https://bugzilla.redhat.com/show_bug.cgi?id=1190070

File: assets/app/views/settings.html [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+      <h1>Project Settings</h1>
+      <alerts alerts="alerts"></alerts>
+      <h2>General information</h2>
+      <!-- TODO make these fields editable - probably with a cog icon that flips into inline-editing -->
+      <dl class="dl-horizontal left">
+        <dt>Display name:</dt>
+        <dd ng-if="project.displayName">{{project.displayName}}</dd>
+        <dd ng-if="!project.displayName"><em>No display name</em></dd>
+        <dt>Description:</dt>
+        <dd ng-if="project.metadata.annotations && project.metadata.annotations.description">{{project.metadata.annotations.description}}</dd>
```
Comment (liggitt) [inline assets/app/views/settings.html]: is `metadata.annotations.description` a standard annotation? If so, should we make a filter to do this for us?

File: assets/app/views/settings.html [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+          <tr ng-if='!quota.status.used' class="danger">
+            <td colspan="5">
+              <span class="pficon-layered" data-toggle="tooltip" data-placement="right" title="Missing quota status" style="cursor: help;">
+                <span class="pficon pficon-error-octagon"></span>
+                <span class="pficon pficon-error-exclamation"></span>
+              </span>
+              Status has not been reported on this quota usage record.  Any resources limited by this quota record can not be allocated.
+            </td>
+          </tr>
+          <tr ng-class="{warning: quota.status.used && quota.status.used[resourceType] >= quota.status.hard[resourceType]}" ng-repeat="(resourceType, specMax) in quota.spec.hard">
```
Comment (liggitt) [inline assets/app/views/settings.html]: where is `resourceType` coming from?

Comment (jwforres) [inline assets/app/views/settings.html]: from ng-repeat="(resourceType, specMax) in quota.spec.hard   you can use variables set from a repeat on the same tag as the repeat

File: assets/app/views/_project-nav.html [hunk 5]
```diff
@@ -11,7 +11,7 @@
         <div class="row">
           <div class="col-md-6">
             <div class="form-group">
-              <label class="control-label">Filter by labels</label>
+              <label ng-if="!renderOptions.hideFilterWidget" class="control-label">Filter by labels</label>
```
Comment (liggitt) [inline assets/app/views/_project-nav.html]: tolerate undefined renderOptions?

File: assets/app/scripts/controllers/settings.js [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+      console.log("quotas", $scope.quotas);
+    });
+
+    DataService.list("limitRanges", $scope, function(limitRanges) {
+      $scope.limitRanges = limitRanges.by("metadata.name");
+      $scope.emptyMessageLimitRanges = "There are no resource limits set on this project.";
+      console.log("limitRanges", $scope.limitRanges);
+    });    
+
+    $scope.$watch("quotas", function() {
```
Comment (liggitt) [inline assets/app/scripts/controllers/settings.js]: Is there a cleaner way to trigger this? Maybe using a directive to render tooltips?

File: assets/app/scripts/directives/nav.js [hunk 7]
```diff
@@ -57,14 +53,18 @@ angular.module('openshiftConsole')
...
-        LabelFilter.setupFilterWidget($(".navbar-filter-widget", element), $(".active-filters", element));
-
         $scope.$watch("projects", function(projects) {
           select.empty();
           updateOptions(projects);
           select.selectpicker('refresh');
         });
+
+        LabelFilter.setupFilterWidget($(".navbar-filter-widget", element), $(".active-filters", element));
+        LabelFilter.toggleFilterWidget(!$scope.hideFilterWidget);
```
Comment (liggitt) [inline assets/app/scripts/directives/nav.js]: `$scope.renderOptions.hideFilterWidget`? Tolerate undefined `renderOptions`?

File: assets/app/scripts/directives/nav.js [hunk 8]
```diff
@@ -57,14 +53,18 @@ angular.module('openshiftConsole')
...
+
+        $scope.$watch("renderOptions", function(renderOptions) {
+          LabelFilter.toggleFilterWidget(!renderOptions.hideFilterWidget);
```
Comment (liggitt) [inline assets/app/scripts/directives/nav.js]: Tolerate undefined `renderOptions`?

File: assets/app/scripts/controllers/settings.js (hunk 6 above)
Comment (jwforres) [inline assets/app/scripts/controllers/settings.js]: Actually yeah that would work, can do an "A" (attribute) style directive

**Comment by liggitt** [comment]
given the data from the API, I guess the units stuff is the best we can do

File: assets/app/scripts/filters/util.js [hunk 9]
```diff
@@ -3,4 +3,28 @@ angular.module('openshiftConsole')
     return function(hash) {
       return Object.keys(hash).length;
     };
+  })
+  .filter('usageWithUnits', function() {
+    return function(value, type) {
+      // only special casing memory at the moment
+      var split = /([0-9]+)\s*(.*)/.exec(value);
```
Comment (liggitt) [inline assets/app/scripts/filters/util.js]: will the amount ever include a decimal?

File: assets/app/views/settings.html [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+      <div ng-if="(limitRanges | hashSize) == 0"><em>{{emptyMessageLimitRanges}}</em></div>
+      <table ng-if="(limitRanges | hashSize) > 0" class="table">
+        <thead>
+          <th>Resource type</th>
+          <th>Min</th>
+          <th>Max</th>
+        </thead>
+        <tbody ng-repeat="limitRange in limitRanges">
+          <tr ng-repeat-start="limit in limitRange.spec.limits"></tr>
+          <tr ng-repeat="(type, min) in limit.min">
```
Comment (liggitt) [inline assets/app/views/settings.html]: if min and max are not guaranteed to have the same set of resources, iterate over max to make sure we surface all max limits

File: assets/app/views/settings.html [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+            <!-- TODO prettify type -->
+            <td>{{limit.type}} {{type}}</td>
+            <!-- TODO prettify min based on the type? -->
+            <td>{{min | usageWithUnits : type}}</td>
```
Comment (liggitt) [inline assets/app/views/settings.html]: if min and max are not guaranteed to have the same set of resources, handle the case where no min is set for a type

**Comment by liggitt** [comment]
LGTM pending min/max question

**Comment by jwforres** [comment]
[merge]

**Comment by jwforres** [comment]
unrelated failure [merge]

**Pull request merged**